aips/aip-20.md

```diff
@@ -0,0 +1,399 @@
+---
+aip: 20
+title: Generic Operations of Algebraic Structures
+author: zhoujun-ma, alin
+discussions-to (*optional): https://github.com/aptos-foundation/AIPs/issues/94
+Status: Draft
+type: Standard (Framework)
+created: 2023/03/15
+---
+
+# AIP-14 - Generic Operations of Algebraic Structures
+## Summary
+
+This AIP proposes some generic functions and structs for performing operations on cryptographic algebraic structures in Move.
+
+The initial list of supported generic operations includes group/field element serialization/deserialization, basic arithmetic, pairing, hash-to-structure, casting.
+
+The list of supported groups starts with groups/fields used in BLS12-381, a popular pairing-friendly curve as described [here](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-11#name-bls-curves-for-the-128-bit-).
+
+Either the operation list or the structure list can be extended by future AIPs.
+
+## Motivation
+
+Algebraic structures are fundamental building blocks for many cryptographic schemes, but also hard to implement efficiently in pure Move.
+This change should allow Move developers to implement generic constructions of those schemes, then get different instantiations by only switching the type parameter(s).
+
+For example, if BLS12-381 groups and BN254 groups are supported, one can implement a generic Groth16 proof verifier construction, then be able to use both BLS12-381-based Groth16 proof verifier and BN254-based Groth16 proof verifier.
+
+BLS12-381-based Groth16 proof verifier has been implemented this way as part of the reference implementation.
+
+## Rationale
+
+An alternative non-generic approach is to expose instantiated schemes directly in stdlib.
+For example, we can define a Groth16 proof verification function
+`0x1::groth16_b::verify_proof(vk, proof, public_inputs): bool`
+for every bilinear map `b`;
+for ECDSA signatures which requires a hash function and a group, we can define
+`0x1::ecdsa_h_g::verify_signature(pk, msg, sig):bool`
+for each pair of proper hash function `h` and group `g`.
+
+Compared with the proposed approach, the alternative approach saves the work of constructing the schemes for Move developers. However, the size of stdlib can multiply too fast in the future.
+
+To keep the Aptos stdlib concise while still covering as many use cases as possible, the proposed generic approach should be chosen over the alternative approach.
+
+## Specifications
+
+### Generic Operations
+
+#### Structs and Functions
+
+Module `aptos_std::algebra` is designed to have the following definitions.
+
+- A generic struct `Element<S>` that represents an element of algebraic structure `S`.
+- Generic functions that represent group/field operations.
+
+Below is the full specification in pseudo-Move.
+
+```rust
+module aptos_std::algebra {
```

Comment:
Can we give this a more specific name than 'algebra'? It sounds weird for some one with math background, because there are many different algebras, and they would be named according to what they do. It could be `crypt_algebra`, or whatever is adequate for this particular instance here.

Reply:
Fixed.